toolset/databases/abstract_database.py

```diff
@@ -0,0 +1,95 @@
+import abc
+import os
+
+class AbstractDatabase:
+    '''
+    Abstract Database Class.
+    To be derived for defining a new concrete Database type
+    '''
+    margin = 1
+    
+    @classmethod
+    @abc.abstractmethod
+    def get_connection(cls, config):
+        '''
+        Establishes and returns a connection to the database.
+        '''
+        pass
+
+    @classmethod
+    @abc.abstractmethod
+    def get_current_world_table(cls, config):
+        '''
+        Returns a JSON object containing all 10,000 World items as they currently
+        exist in the database. This is used for verifying that entries in the
+        database have actually changed during an Update verification test.
+        '''
+        pass
+
+    @classmethod
+    @abc.abstractmethod
+    def test_connection(cls, config):
+        '''
+        Tests the connection and returns true if it is established.
+        '''
+        pass
+
+    @classmethod
+    @abc.abstractmethod
+    def get_queries(cls, config):
+        '''
+        Returns the number of db queries (all types).
+        '''
+        pass
+
+    @classmethod
+    @abc.abstractmethod
+    def get_rows(cls, config):
+        '''
+        Returns the number of db rows read.
+        '''
+        pass
+
+    @classmethod
+    @abc.abstractmethod
+    def get_rows_updated(cls, config):
+        '''
+        Return the number of updated db rows.
+        '''
+        pass
+
+    @classmethod
+    @abc.abstractmethod
+    def reset_cache(cls, config):
+        '''
+        Reset the query cache.
+        '''
+        pass
+
+    @classmethod
+    def verify_queries(cls, config, table_name, url, concurrency=512, count=2, check_updates=False):
+        '''
+        Verify query and row numbers for table_name.
+        Retrieve from the database statistics of the number of queries made, the number of rows read, eventually the number of updated rows.
+        Run 2 repetitions of http requests at the concurrency level 512 with siege.
+        Retrieve statistics again, calculate the number of queries made and the number of rows read.
+        '''
+        rows_updated = None
+        cls.tbl_name = table_name # used for Postgres and mongodb
+
+        queries = int(cls.get_queries(config))
+        rows = int(cls.get_rows(config))
+        if check_updates:
+            rows_updated = int(cls.get_rows_updated(config))
+
+        cls.reset_cache(config)
+        #Start siege requests
+        path = config.db_root
+        os.system("siege -c %s -r %s %s -R %s/.siegerc" % (concurrency, count, url, path))
```

Comment:
How is this better than making a single HTTP request to the server?

Is the purpose of this to identify implementations which are making good-faith attempts at queries, but failing [somehow] and therefore getting better, albeit fake, performance?

Reply:
> How is this better than making a single HTTP request to the server?

It's a bit of an arbitrary choice,
maybe in case the second repetition reveals something other than the first?
`repetition` is an adjustable variable.

> Is the purpose of this to identify implementations which are making good-faith attempts at queries, but failing [somehow] and therefore getting better, albeit fake, performance?

Yes, what matters is to identify the non-compliance with the test specifications, there is an obligation of means, but also an obligation of result, whatever the performance (better or not).


For the moment, I have identified three possible cases causing failures:
- non-compliance with the code (for example, updating without prior reading)
- failed Http requests
- connections to the database probably impossible due to overload (sometimes causing errors, but not always)

We can possibly add stats on the number of `connections` to the database, to see if it provides answers.